## 1.1. Package imports

In [1]:
import pandas as pd

----
## 1.2. Display settings

In [2]:
pd.set_option("display.width", 140)

----
## 1.3. Reading in the data

In [3]:
df = pd.read_csv("../data/raw/Retail data.csv", sep=";")
# df = pd.read_csv("../data/raw/Potential Customers.csv", sep=";")

----
## 1.4. Exploring data features
This step involves looking at the data without altering it.

In [4]:
# First 5 rows
df.head()

Cocunut Mortgage_YN  AGE_AT_ORIGINATION  AGE  YEARS_WITH_BANK MARTIAL_STATUS EDUCATION EMPLOYMENT GENDER  CUST_INCOME  \
0        1           Y                50.0   52               13              M       HGH        PVE      M  909,5013077   
1        9           Y                48.0   49               11              M       HGH        SFE      M  288,4615385   
2       11           Y                53.0   55               14              M       BCR        STE      M  1280,528692   
3       12           Y                64.0   66               10              M       BCR        OTH      F  620,9597692   
4       18           Y                46.0   47                9              S       MAS        PVE      F  2239,853846   

  CURRENT_ADDRESS_DATE CURRENT_JOB_DATE CURRENT_WITH_BANK_DATE CURRENT_BALANCE_EUR  
0           1979-04-18       2010-01-01             2004-03-16         7648,350692  
1           1999-12-16       2004-01-20             2005-11-07         30189,90492  
2           2005-06-28       1990-09-01             2003-06-25         50553,17454  
3           2000-12-22       9999-10-01             2006-12-21         15907,28338  
4           2009-08-01       2013-12-24             2007-08-07         27916,19262

In [5]:
# Last 5 rows
df.tail()

Cocunut Mortgage_YN  AGE_AT_ORIGINATION  AGE  YEARS_WITH_BANK MARTIAL_STATUS EDUCATION EMPLOYMENT GENDER  CUST_INCOME  \
23978    79979           N                 NaN   67                3              M       BCR        RET      M  179,2163077   
23979    79982           N                 NaN   59               13              M       HGH        PVE      F  690,0951538   
23980    79983           N                 NaN   68                7              W       PRS        RET      M  255,9498462   
23981    79985           N                 NaN   59                2              M       BCR        STE      M  459,3230769   
23982    79998           N                 NaN   47                1              M       HGH        STE      F  528,7692308   

      CURRENT_ADDRESS_DATE CURRENT_JOB_DATE CURRENT_WITH_BANK_DATE CURRENT_BALANCE_EUR  
23978           1999-03-10       9999-10-01             2014-06-02              781,19  
23979           1997-01-01       2004-01-01             2004-01-12         13716,36638  
23980           1949-10-27       9999-10-01             2009-12-01             1066,28  
23981           2003-04-01       2008-09-24             2015-04-23         4577,400385  
23982           2010-01-01       2005-03-23             2016-06-06         2307,692308

In [6]:
# Numeric column statistics
df.describe().T

count          mean           std   min      25%      50%       75%      max
Cocunut             23983.0  39969.697786  22996.273034   1.0  20076.0  40066.0  59778.00  79998.0
AGE_AT_ORIGINATION    306.0     41.333333      8.970992  22.0     35.0     40.0     47.75     64.0
AGE                 23983.0     49.004378     12.976205  20.0     39.0     49.0     60.00     92.0
YEARS_WITH_BANK     23983.0      7.370804      4.320568   0.0      3.0      9.0     11.00     40.0

In [7]:
# Non-numeric column statistics
df.describe(include=[object]).T

count unique          top   freq
Mortgage_YN             23983      2            N  23677
MARTIAL_STATUS          23983      5            M  17024
EDUCATION               23983      8          HGH  15987
EMPLOYMENT              23983      5          PVE  10743
GENDER                  23983      2            F  12105
CUST_INCOME             23983  18823  230,7692308    217
CURRENT_ADDRESS_DATE    23983   9000   2000-01-01    370
CURRENT_JOB_DATE        23983   5645   9999-10-01   5771
CURRENT_WITH_BANK_DATE  23983   4142   2005-07-26     47
CURRENT_BALANCE_EUR     23983  18415            0   2911

In [8]:
# Feature overview
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23983 entries, 0 to 23982
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Cocunut                 23983 non-null  int64  
 1   Mortgage_YN             23983 non-null  object 
 2   AGE_AT_ORIGINATION      306 non-null    float64
 3   AGE                     23983 non-null  int64  
 4   YEARS_WITH_BANK         23983 non-null  int64  
 5   MARTIAL_STATUS          23983 non-null  object 
 6   EDUCATION               23983 non-null  object 
 7   EMPLOYMENT              23983 non-null  object 
 8   GENDER                  23983 non-null  object 
 9   CUST_INCOME             23983 non-null  object 
 10  CURRENT_ADDRESS_DATE    23983 non-null  object 
 11  CURRENT_JOB_DATE        23983 non-null  object 
 12  CURRENT_WITH_BANK_DATE  23983 non-null  object 
 13  CURRENT_BALANCE_EUR     23983 non-null  object 
dtypes: float64(1), int64(3), object(10)
me

In [9]:
# Number of missing values in each column
df.isna().sum()

Cocunut                       0
Mortgage_YN                   0
AGE_AT_ORIGINATION        23677
AGE                           0
YEARS_WITH_BANK               0
MARTIAL_STATUS                0
EDUCATION                     0
EMPLOYMENT                    0
GENDER                        0
CUST_INCOME                   0
CURRENT_ADDRESS_DATE          0
CURRENT_JOB_DATE              0
CURRENT_WITH_BANK_DATE        0
CURRENT_BALANCE_EUR           0
dtype: int64

In [10]:
# Number of unique values in each column
df.nunique()

Cocunut                   23983
Mortgage_YN                   2
AGE_AT_ORIGINATION           41
AGE                          73
YEARS_WITH_BANK              18
MARTIAL_STATUS                5
EDUCATION                     8
EMPLOYMENT                    5
GENDER                        2
CUST_INCOME               18823
CURRENT_ADDRESS_DATE       9000
CURRENT_JOB_DATE           5645
CURRENT_WITH_BANK_DATE     4142
CURRENT_BALANCE_EUR       18415
dtype: int64

In [11]:
# Unique categories
for column in ["Mortgage_YN", "MARTIAL_STATUS", "EDUCATION", "EMPLOYMENT", "GENDER"]:
    print(f"{column}: {df[column].unique().tolist()}")

Mortgage_YN: ['Y', 'N']
MARTIAL_STATUS: ['M', 'S', '*noval*', 'D', 'W']
EDUCATION: ['HGH', 'BCR', 'MAS', 'SEC', 'PHD', 'PRS', 'PRI', 'OTH']
EMPLOYMENT: ['PVE', 'SFE', 'STE', 'OTH', 'RET']
GENDER: ['M', 'F']


----
## 1.5. Retail data notes
* Redundant column: Cocunut
* CUST_INCOME and CURRENT_BALANCE_EUR are incorrectly parsed due to comma decimal separator
* Date columns: CURRENT_ADDRESS_DATE, CURRENT_JOB_DATE, CURRENT_WITH_BANK_DATE
* Invalid date: 9999-10-01
* Missing AGE_AT_ORIGINATION values: 23677
* MARTIAL_STATUS has \*noval\* category

----
## 1.6. Potential customer data notes
* Similar issues as with retail data
* Column differences (missing): Mortgage_YN, AGE_AT_ORIGINATION

----
## 1.7. Column descriptions

1. **AGE**: Age of customer (`int`)
2. **YEARS_WITH_BANK**: Number of years the customer has been with the bank (`int`)
3. **MARTIAL_STATUS**: Customer family state (`category`)
4. **EDUCATION**: Customer education level (`category`)
5. **EMPLOYMENT**: Customer emplyment status (`category`)
6. **GENDER**: Female or male (`category`)
7. **CUST_INCOME**: Customer monthly income in EUR (`float`)
8. **CURRENT_ADDRESS_DATE**: Date when customer moved to their current address (`date`)
9. **CURRENT_JOB_DATE**: Date when customer started their current job (`date`)
10. **CURRENT_WITH_BANK_DATE**: Date when customer opened their current bank account (`date`)
11. **CURRENT_BALANCE_EUR**: Current balance of customer in EUR (`float`)

 

12. **Mortgage_YN**: Y if customer has a mortgage, N if not (`category`)
13. **AGE_AT_ORIGINATION**: Age of customer when they took the mortgage (`int`)